### ISSUE TITLE
Provide an equivalent of curio.monitor

### ISSUE THREAD

**Issue by Nikratio** [python-trio/trio#413]
curio.monitor allows to debug an async program while it's running. It is fantastic for debugging, and it would be great to have something like this for trio.

**Comment by njsmith** [comment]
Yeah, it would! I'm surprised we don't have an issue open for this already.

For those who haven't seen it, [there's some monitor examples here](https://curio.readthedocs.io/en/latest/tutorial.html#tasks), and [also check this out](https://twitter.com/swfiua/status/841465029409017856).

I do think we should use a slightly different architecture: in curio the monitor is a single piece of monolithic code that's built into the core library, has a hardcoded mechanism for connecting, and has a small fixed set of commands, which are implemented using internal, non-public APIs. It's an awesome and useful proof-of-concept/demo, and we should provide a similar experience, but I'd like to do it in a way that lets monitor implementations evolve independently from the core trio library.

A lot of the groundwork for this is already there: Trio provides lots of introspection APIs. (Probably there are some gaps, but we can sort that out as we go.) For example, [here's a little experiment in printing nice task trees from a live program](https://github.com/python-trio/trio/blob/master/notes-to-self/print-task-tree.py), and there are lots of other things you can do – show [task tracebacks](https://trio.readthedocs.io/en/latest/reference-hazmat.html#trio.hazmat.Task.coro), visualizing [execution traces](https://trio.readthedocs.io/en/latest/tutorial.html#task-switching-illustrated), etc. So I think the next step here would be for someone -- could be anyone -- to start building something and see how it goes. The [curio monitor implementation is not very big](https://github.com/dabeaz/curio/blob/master/curio/monitor.py).

In the longer run I think it'd be really neat to have a monitor interface that's an IPython REPL + magic commands for stuff like introspecting the task tree. But that's going to take some non-trivial work in IPython and python-prompt-toolkit to make them async/trio friendly (CC @Carreau).

**Comment by Carreau** [comment]
IIRC prompt_toolit is already quite async-await friendly. IPython less so, but I'd love to convert all internal IPython API to be sans IO or async await compatible.

**Comment by tgoodlet** [comment]
>I think it'd be really neat to have a monitor interface that's an IPython REPL + magic commands for stuff like introspecting the task tree.

👍 👍 👍 that's exactly how it should go :)

**Comment by touilleMan** [comment]
> So I think the next step here would be for someone -- could be anyone -- 

I seems this anyone is *me* :smile:

**Comment by lithp** [comment]
Does the existence of https://github.com/python-trio/trio-monitor mean this issue can be closed?

**Comment by njsmith** [comment]
That repo is pretty much empty right now, unfortunately.

**Comment by Tronic** [comment]
There are three older (somewhat dead) monitoring projects that I know of, and my concept test from earlier this week https://github.com/Tronic/trio-web-monitor in case anyone is interested.

What is needed is someone willing to maintain the project long term, and then perhaps take the best elements of each of the existing implementations (then it is already a lot better than curio monitor or any other Python debugger for that matter). I am willing to participate if someone else shows up to cooperate but I won't take sole responsibility over it.
